postgis/gserialized_supportfn.c

```diff
@@ -0,0 +1,372 @@
+/**********************************************************************
+ *
+ * PostGIS - Spatial Types for PostgreSQL
+ * http://postgis.net
+ *
+ * PostGIS is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * PostGIS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************/
+
+
+#include "../postgis_config.h"
+
+#if POSTGIS_PGSQL_VERSION >= 120
+
+/* PostgreSQL */
+#include "postgres.h"
+#include "funcapi.h"
+#include "access/htup_details.h"
+#include "access/stratnum.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_opfamily.h"
+#include "catalog/pg_type_d.h"
+#include "catalog/pg_am_d.h"
+#include "nodes/supportnodes.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/makefuncs.h"
+#include "optimizer/optimizer.h"
+#include "parser/parse_func.h"
+#include "utils/array.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/numeric.h"
+#include "utils/syscache.h"
+
+/* PostGIS */
+#include "liblwgeom.h"
+
+/* Prototypes */
+Datum postgis_index_supportfn(PG_FUNCTION_ARGS);
+
+
+/*
+* Depending on the function, we will deploy different
+* index enhancement strategies. Containment functions
+* can use a more strict index strategy than overlapping
+* functions. For within-distance functions, we need
+* to construct expanded boxes, on the non-indexed
+* function argument. We store the metadata to drive
+* these choices in the IndexableFunctions array.
+*/
+typedef struct
+{
+	const char *fn_name;
+	int   strategy_number; /* Index strategy to add */
+	int   nargs;           /* Expected number of function arguments */
+	int   expand_arg;      /* Radius argument for "within distance" search */
+} IndexableFunction;
+
+/*
+* Metadata currently scanned from start to back,
+* so most common functions first. Could be sorted
+* and searched with binary search.
```

Comment:
what if we make many support functions, for each original - will it be even faster without a check?

Reply:
We could make one per strategy, but honestly it's not worth it. The function is called once at the planning stage, so this wee bit of inefficiency is not an issue.